docs/contributing/MODEL_MANAGER.md

```diff
@@ -0,0 +1,1214 @@
+| `event_handlers` | List[DownloadEventHandler]|     | Event handlers (see below)        |
+| `job_started`    | float            |              | Timestamp for when the job started running |
+| `job_ended`      | float            |              | Timestamp for when the job completed or errored out |
+| `job_sequence`   | int              |              | A counter that is incremented each time a model is dequeued |
+| `preserve_partial_downloads`| bool  | False        | Resume partial downloads when relaunched.   |
+| `error`          | Exception        |              | A copy of the Exception that caused an error during download |
+
+When you create a job, you can assign it a `priority`. If multiple
+jobs are queued, the job with the lowest priority runs first. (Don't
+blame me! The Unix developers came up with this convention.) 
+
+Every job has a `source` and a `destination`. `source` is an untyped
+object in the base class, but subclassses redefine it more
+specifically.
+
+The `destination` must be the Path to a file or directory on the local
+filesystem. If the Path points to a new or existing file, then the
+source will be stored under that filename. If the Path ponts to an
+existing directory, then the downloaded file will be stored inside the
+directory, usually using the name assigned to it at the remote site in
+the `content-disposition` http field.
+
+When the job is submitted, it is assigned a numeric `id`. The id can
+then be used to control the job, such as starting, stopping and
+cancelling its download.
+
+The `status` field is updated by the queue to indicate where the job
+is in its lifecycle. Values are defined in the string enum
+`DownloadJobStatus`, a symbol available from
+`invokeai.app.services.download_manager`. Possible values are:
+
+| **Value**    |   **String Value**  | ** Description ** |
+|--------------|---------------------|-------------------|
+| `IDLE`         | idle              | Job created, but not submitted to the queue |
+| `ENQUEUED`     | enqueued          | Job is patiently waiting on the queue       |
+| `RUNNING`      | running           | Job is running!                             |
+| `PAUSED`       | paused            | Job was paused and can be restarted         |
+| `COMPLETED`    | completed         | Job has finished its work without an error |
+| `ERROR`        | error             | Job encountered an error and will not run again|
+| `CANCELLED`    | cancelled         | Job was cancelled and will not run (again) |
+
+`job_started`, `job_ended` and `job_sequence` indicate when the job
+was started (using a python timestamp), when it completed, and the
+order in which it was taken off the queue. These are mostly used for
+debugging and performance testing.
+
+In case of an error, the Exception that caused the error will be
+placed in the `error` field, and the job's status will be set to
+`DownloadJobStatus.ERROR`. 
```

Comment:
The code currently uses a `tempfile.TemporaryDirectory` to store downloads in progress. If the app is shutdown cleanly, then the directory and its contents will be deleted. If the app crashes or is killed in a way that doesn't let Python run its context exit routine, then the temporary directory will still be there on restart but ignored.

Here are a few alternatives of increasing complexity. Which do you prefer?

1. On startup, the system scans for dangling temporary directories and deletes them and their contents.
2. Instead of storing the temporary directory under the randomly-assigned name that `TemporaryDirectory` gives it, the name is assigned deterministically using the download source URL and doesn't get deleted. This allows the user to restart the download and the download will resume where it left off using the partial files. (This is actually an easy code change since restart of partial downloads is already implemented)
3. Some state information about downloads in progress are written into the invokeai database, allowing all downloads to be resumed automatically at startup time.